kernel/comps/virtio/src/device/block/device.rs

```diff
@@ -175,104 +151,177 @@ impl DeviceInner {
         Ok(device)
     }
 
-    /// Reads data from the block device, this function is blocking.
-    /// FIEME: replace slice with a more secure data structure to use dma mapping.
-    pub fn read(&self, sector_id: Sid, buf: &[VmWriter]) {
-        // FIXME: Handling cases without id.
-        let id = self.id_allocator.lock().pop().unwrap() as usize;
-        let req = BlockReq {
-            type_: ReqType::In as _,
-            reserved: 0,
-            sector: sector_id.to_raw(),
+    /// Handles the irq issued from the device
+    fn do_handle_irq(&self) {
+        loop {
+            // Pops the complete request
+            let complete_request = {
+                let mut queue = self.queue.lock_irq_disabled();
+                let Ok((token, _)) = queue.pop_used() else {
+                    return;
+                };
+                self.submitted_requests.lock().remove(&token).unwrap()
+            };
+
+            // Handles the response
+            complete_request.resp.sync().unwrap();
+            let resp: BlockResp = complete_request.resp.read_val(0).unwrap();
+            self.id_allocator.lock().free(complete_request.id as usize);
+            match RespStatus::try_from(resp.status).unwrap() {
+                RespStatus::Ok => {}
+                _ => panic!("io error in block device"),
+            };
```

Comment:
We should return an I/O error instead of triggering a kernel panic.

Reply:
Add a `FIXME` here to fix this in the upcoming commit.